### ISSUE TITLE
Proxy timeout from Nginx after 90 seconds when filtering users to undelete

### ISSUE THREAD

**Issue by d0dge** [NodeBB/nodebb-plugin-undelete-users#2]
If I try and filter deleted users by email address, it tries for 90 seconds and then Nginx think the connection has timed out:

`upstream timed out (110: Connection timed out) while reading response header from upstream`

I can increase the timeout of course, but I would have thought 90 seconds should be enough to find a user.  We probably have around 2000 deleted users.

Is this expected?

**Comment by d0dge** [comment]
If I search by email, I get the following error:

Oops! Looks like something went wrong!
Cannot read properties of undefined (reading 'toLowerCase')

No hint in the logs as to which line of code this is being caused by, unfortunately.

Basically, the search within this plugin doesn't work at all.

This is on nodebb v3.2.1.

**Comment by barisusakli** [comment]
If you open up logs/output.log and find that error it should have a stack trace.

**Comment by barisusakli** [comment]
Released 1.0.2, let me know if you still get that error

**Comment by d0dge** [comment]
> If you open up logs/output.log and find that error it should have a stack trace.

That's what I meant.... nothing written to output.log when the issue occurred.

I don't get the "toLowerCase" issue any more after upgrading to 1.0.2. but searches still timeout after 90 seconds.

**Comment by barisusakli** [comment]
Can you put a `console.log(page)` after this line https://github.com/NodeBB/nodebb-plugin-undelete-users/blob/main/lib/controllers.js#L30, and then do a search. Then check what happens in the nodebb logs. Since you have 2000 undelete events it should call that line at least 66 times if the email you are searching is not in the list.

You can see logs by running `./nodebb log`  after starting nodebb.

**Comment by d0dge** [comment]
I searched for one of the last few accounts that got deleted.  It's run for probably an hour now (and still going) and below is the output.  There are anywhere between 1 and 5 seconds between each line.

`1
2
1
2
3
4
5
6
7
8
9
10
11
12
13
1
2
3
14
4
5
6
15
7
8
16
9
10
17
11
12
18
13
19
14
1
2
3
15
4
20
5
16
6
17
7
1
2
3
8
21
4
18
5
9
6
10
7
22
19
8
11
9
12
20
10
23
13
11
21
12
14
24
13
15
22
14
25
16
15
23
17
16
26
18
24
17
27
19
18
25
20
19
28
26
21
20
29
27
22
21
22
23
30
28
23
24
29
31
24
25
30
32
25
26
31
26
33
27
27
32
28
34
28
33
29
35
29
34
30
30
36
35
31
31
37
32
36
32
38
33
37
33
39
34
38
34
35
40
35
39
36
36
41
40
37
37
42
41
38
38
43
42
39
39
43
44
40
40
44
45
41
41
45
46
42
42
46
47
43
43
47
44
48
44
48
45
49
45
`

**Comment by barisusakli** [comment]
Hmm that seems a bit too much, what does this output show in mongodb `db.objects.stats()`

**Comment by d0dge** [comment]
We're on Redis.  This is the memory and stats output from INFO, is that the equivalent to what you requested?

```
# Memory
used_memory:1082473408
used_memory_human:1.01G
used_memory_rss:1098792960
used_memory_rss_human:1.02G
used_memory_peak:1598416360
used_memory_peak_human:1.49G
used_memory_peak_perc:67.72%
used_memory_overhead:80967686
used_memory_startup:791000
used_memory_dataset:1001505722
used_memory_dataset_perc:92.59%
allocator_allocated:1082464328
allocator_active:1504980992
allocator_resident:1536208896
total_system_memory:5998456832
total_system_memory_human:5.59G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.39
allocator_frag_bytes:422516664
allocator_rss_ratio:1.02
allocator_rss_bytes:31227904
rss_overhead_ratio:0.72
rss_overhead_bytes:-437415936
mem_fragmentation_ratio:1.02
mem_fragmentation_bytes:16361552
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:318302
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0

# Stats
total_connections_received:1033
total_commands_processed:10597855069
instantaneous_ops_per_sec:187089
total_net_input_bytes:752142193070
total_net_output_bytes:136024558808
instantaneous_input_kbps:8287.44
instantaneous_output_kbps:731.31
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:1646567
expired_stale_perc:0.00
expired_time_cap_reached_count:0
evicted_keys:0
keyspace_hits:1252944456
keyspace_misses:8699723473
pubsub_channels:5
pubsub_patterns:2
latest_fork_usec:38223
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

```

**Comment by barisusakli** [comment]
That's fine I was wondering if you were on mongodb and didn't have the indices. It seems the way pagination is implemented in this plugin is the problem.

It keeps growing a list of events on each iteration and calling the same methods with bigger list everytime. This would cause each iteration to slow down. Seems like it will need a refactor to work faster.

**Comment by barisusakli** [comment]
Try 1.0.4 and see if it loads faster when you filter by email

**Comment by d0dge** [comment]
This works much better.

It doesn't exactly pick out the specific user I search for, but it navigates to the page of deleted users which has that user, so I can see them on the page.  This works for me :)

**Comment by barisusakli** [comment]
Hmm it should actually show only the user with the matching email, does it show users with no email when you search by email

**Comment by barisusakli** [comment]
Published 1.0.5, let me know if you still get users that don't match the email address. Gonna close this as the perf issue is resolved.

**Comment by d0dge** [comment]
Sorry for the delay, I just had chance to upgrade to the new version.

We're just getting the searched user now, thanks.  You were right, the other users that appeared were ones that had no email address linked to them.
